src/libraries/System.Private.CoreLib/src/System/IO/BinaryWriter.cs

```diff
@@ -182,18 +179,39 @@ public virtual void Write(byte[] buffer, int index, int count)
         // advanced by two.
         // Note this method cannot handle surrogates properly in UTF-8.
         //
-        public virtual unsafe void Write(char ch)
+        public virtual void Write(char ch)
         {
-            if (char.IsSurrogate(ch))
+            if (!Rune.TryCreate(ch, out Rune rune)) // optimistically assume UTF-8 code path (which uses Rune) will be hit
+            {
                 throw new ArgumentException(SR.Arg_SurrogatesNotAllowedAsSingleChar);
+            }
+
+            Span<byte> buffer = stackalloc byte[8]; // reasonable guess for worst-case expansion for any arbitrary encoding
 
-            Debug.Assert(_encoding.GetMaxByteCount(1) <= 16, "_encoding.GetMaxByteCount(1) <= 16)");
-            int numBytes = 0;
-            fixed (byte* pBytes = &_buffer[0])
+            if (_useFastUtf8)
+            {
+                int utf8ByteCount = rune.EncodeToUtf8(buffer);
+                OutStream.Write(buffer.Slice(0, utf8ByteCount));
+            }
+            else
             {
-                numBytes = _encoder.GetBytes(&ch, 1, pBytes, _buffer.Length, flush: true);
+                byte[]? rented = null;
+                int maxByteCount = _encoding.GetMaxByteCount(1);
+
+                if (maxByteCount > buffer.Length)
+                {
```

Comment:
is it possible to hit this code path?

Reply:
In theory I could write a custom encoding that expanded a single char to any number of bytes.  That said, I currently care very little for such a scenario.

Reply:
Yes, it's possible to hit this. The encoder used by the XML stack has a worst-case `GetMaxByteCount(1)` return value of __72 bytes__. This is calculated as `(1 + 1) * 12 * 3`, where the 12 comes from [here](https://github.com/dotnet/runtime/blob/8a52f1e948b6f22f418817ec1068f07b8dae2aa5/src/libraries/System.Private.Xml/src/System/Xml/Core/CharEntityEncoderFallback.cs#L35-L41) and the 3 comes from [here](https://github.com/dotnet/runtime/blob/8a52f1e948b6f22f418817ec1068f07b8dae2aa5/src/libraries/System.Private.CoreLib/src/System/Text/UTF8Encoding.cs#L56).